File: trunk/Source/Runtime/CPlusplus/Animation/Bibim/EvalTimeflow.cpp
#include <Bibim/PCH.h>
#include <Bibim/EvalTimeflow.h>
#include <Bibim/ComponentStreamReader.h>
#include <Bibim/Math.h>
#include <Bibim/Timeline.h>

namespace Bibim
{
    BBImplementsComponent(EvalTimeflow);

    EvalTimeflow::EvalTimeflow()
        : timeline(nullptr),
          time(0.0f),
          duration(0.0f),
          velocity(1.0f),
          stopBehavior(PauseAtStop),
          looped(true),
          isUpdating(false)
    {
    }

    EvalTimeflow::~EvalTimeflow()
    {
        if (timeline != nullptr && isUpdating)
            timeline->Remove(this);
    }

    void EvalTimeflow::Update(float dt, int timestamp)
    {
        if (lastTimestamp == timestamp || velocity == 0.0f)
            return;
        else
            lastTimestamp = timestamp;

        const float nextTime = time + (dt * velocity);

        if (0 <= nextTime && nextTime <= duration)
            time = nextTime;
        else
        {
            if (nextTime > duration)
            {
                if (looped && duration > 0.0f)
                    time = Math::Mod(nextTime, duration);
                else
                    time = duration;
            }
            else
            {
                if (looped && duration > 0.0f)
                    time = duration - Math::Mod(-nextTime, duration);
                else
                    time = 0.0f;
            }
        }
    }

    void EvalTimeflow::Start()
    {
        if (isUpdating)
            return;

        isUpdating = true;
        if (timeline != nullptr)
            timeline->Add(this, this);
    }

    void EvalTimeflow::Stop()
    {
        if (isUpdating == false)
            return;

        isUpdating = false;
        if (timeline != nullptr)
            timeline->Remove(this);

        switch (stopBehavior)
        {
            case PauseAtStop:
                break;
            case ResetAtStop:
                time = 0.0f;
                break;
            case FinishAtStop:
                time = duration;
                break;
        }
    }

    void EvalTimeflow::Reset()
    {
        time = 0.0f;
    }

    float EvalTimeflow::Evaluate(EvalContext& /*context*/)
    {
        return time;
    }

    void EvalTimeflow::SetTimeline(Timeline* value)
    {
        if (timeline != value)
        {
            if (timeline != nullptr && isUpdating)
                timeline->Remove(this);

            timeline = value;

            if (timeline != nullptr && isUpdating)
                timeline->Add(this);
        }
    }

    void EvalTimeflow::SetDuration(float value)
    {
        value = Math::Max(value, 0.0f);
    }

    void EvalTimeflow::OnRead(ComponentStreamReader& reader)
    {
        Base::OnRead(reader);
        timeline = static_cast<Timeline*>(reader.ReadModule(Timeline::ClassID));
        time = 0.0f;
        duration = reader.ReadFloat();
        velocity = reader.ReadFloat();
        stopBehavior = static_cast<Behavior>(reader.ReadByte());
        looped = reader.ReadBool();
    }

    void EvalTimeflow::OnCopy(const GameComponent* original, CloningContext& context)
    {
        Base::OnCopy(original, context);
        const This* o = static_cast<const This*>(original);
        timeline = o->timeline;
        time = o->time;
        duration = o->duration;
        velocity = o->velocity;
        stopBehavior = o->stopBehavior;
        looped = o->looped;
        isUpdating = o->isUpdating;

        if (timeline != nullptr && isUpdating)
            timeline->Add(this);
    }
}

File: Source/Runtime/CSharp/Bibim.Runtime/Animation/Evaluation/EvalTimeflow.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Bibim.Animation
{
    [ClassID('e', 'T', 'F', 'f')]
    public sealed class EvalTimeflow : Eval<float>
    {
        #region Fields
        private float duration;
        private int lastTimestamp;
        #endregion

        #region Properties
        public float Time
        {
            get;
            private set;
        }

        public float Duration
        {
            get { return duration; }
            set
            {
                duration = Math.Max(value, 0.0f);
            }
        }

        public float Velocity
        {
            get;
            set;
        }

        public bool IsLooped
        {
            get;
            set;
        }
        #endregion

        #region Constructors
        public EvalTimeflow()
            : this(0.0f, 1.0f, true)
        {
        }

        public EvalTimeflow(float duration)
            : this(duration, 1.0f, true)
        {
        }

        public EvalTimeflow(float duration, float velocity, bool isLooped)
        {
            Time = 0.0f;
            duration = Math.Max(duration, 0.0f);
            Velocity = velocity;
            IsLooped = isLooped;
        }
        #endregion

        #region Methods
        public void Update(float dt, int timestamp)
        {
            if (lastTimestamp == timestamp || Velocity == 0.0f)
                return;
            else
                lastTimestamp = timestamp;

            float nextTime = Time + dt;

            if (0 <= nextTime && nextTime <= duration)
                Time = nextTime;
            else
            {
                if (nextTime < duration)
                {
                    if (IsLooped && duration > 0.0f)
                        Time = nextTime % duration;
                    else
                        Time = duration;
                }
                else
                {
                    if (IsLooped && duration > 0.0f)
                        Time = duration - (-nextTime % duration);
                    else
                        Time = 0.0f;
                }
            }
        }
        #endregion

        public override void Reset()
        {
            Time = 0.0f;
        }

        public override float Evaluate(EvalContext context)
        {
            return Time;
        }
    }
}
